### PR TITLE
Make dispatch of KIR and FusionIR more similar.

### PR DESCRIPTION
**Pull request by csarofeen**
@naoyam for general review, mostly a mechanical PR
@shmsong would you mind checking kir expr eval performance to see if there's any substantial changes?

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/dispatch.h [hunk 1]
```diff
@@ -82,203 +82,87 @@ class ViewOp;
...
-    TORCH_INTERNAL_ASSERT(false, "Handle not overriden for ShiftOp.");
-  }
-  virtual void handle(const GatherOp*) {
-    TORCH_INTERNAL_ASSERT(false, "Handle not overriden for GatherOp.");
-  }
-  virtual void handle(const ViewOp*) {
-    TORCH_INTERNAL_ASSERT(false, "Handle not overriden for ViewOp.");
-  }
+ protected:
+  virtual void unhandled(const Statement* stmt) override;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/dispatch.h]: Nit: should this be `final`?

File: torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp [hunk 2]
```diff
@@ -351,9 +351,9 @@ class ValidatePlacementAfterWrites : private kir::IrVisitor {
   ValidatePlacementAfterWrites(const std::unordered_set<kir::Expr*>& writes)
       : writes_(writes) {}
 
-  void handle(kir::Expr* expr) {
+  void handle(kir::Expr* expr) final {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp:354]: Not a request to this PR, but wouldn't it be convenient if we could define a `handle` only for non-scoping expressions?

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp [hunk 3]
```diff
@@ -493,15 +493,15 @@ void ExprSegmentationSorter::resetLevels() {
   std::vector<ExprGroup*> next_to_visit;
 
   while (!to_visit_.empty()) {
-    auto visit = to_visit_.front();
+    auto handle = to_visit_.front();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Do you really want to rename this variable? Or is it just accidentally done together with the dispatch functions?

**Comment by shmsong** [comment]
I don't see noticeable perf change on the hot path. 

```cpp
commit dada9aa1c308d7bf614e4094ea6b464eae81cc85 (HEAD -> devel, origin/devel, origin/HEAD)
Author: jjsjann123 <jiej@nvidia.com>
Date:   Sun Dec 12 16:08:10 2021 -0800

-------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations
-------------------------------------------------------------------------------------------
LayerNormBackward_ShapeInference                       56.9 us         56.9 us        12170
LayerNormForward_ShapeInference                        13.3 us         13.3 us        52451
LayerNormBackward_NoShapeInferenceCachedBaseline       29.2 us         29.2 us        24154
LayerNormForward_NoShapeInferenceCachedBaseline        5.52 us         5.52 us       127602



commit 7fe798106bb0a3e9fb94e5e7f8e6e7a68347608d (HEAD, origin/kir_cleanup)
Author: Christian Sarofeen <csarofeen@nvidia.com>
Date:   Mon Dec 13 12:44:35 2021 -0500

    Convert KIR to a Fusion IR Dispatch-like system

-------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations
-------------------------------------------------------------------------------------------
LayerNormBackward_ShapeInference                       55.2 us         55.2 us        12816
LayerNormForward_ShapeInference                        14.0 us         14.0 us        50351
LayerNormBackward_NoShapeInferenceCachedBaseline       27.3 us         27.3 us        25144
LayerNormForward_NoShapeInferenceCachedBaseline        5.81 us         5.81 us       119682


```

File: torch/csrc/jit/codegen/cuda/dispatch.cpp [hunk 4]
```diff
@@ -386,46 +367,165 @@ void OptOutConstDispatch::handle(const Val* v) {
...
 }
 
-Statement* OptOutMutator::mutate(Val* v) {
-  // If value is already mutated, return the mutation
-  if (mutations.find(v) != mutations.end())
-    return mutations[v];
-  return Val::mutatorDispatch(this, v);
+// Vals
+void OptOutConstDispatch::handle(const IterDomain* stmt) {
+  unhandled(stmt->as<Statement>());
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/dispatch.cpp]: Does it need `->as<Statement>()`? Build seems fine without it.

File: torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  virtual void handle(GridBroadcast* stmt);
+  virtual void handle(GridWelford* stmt);
+};
+
+class TORCH_CUDA_CU_API OptInConstDispatch : public OptOutConstDispatch {
+ public:
+  using OptOutConstDispatch::handle;
+
+ protected:
+  virtual void unhandled(const Node* stmt) override;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h]: Shouldn't this be `final`?

File: torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+};
+
+class TORCH_CUDA_CU_API OptInDispatch : public OptOutDispatch {
+ public:
+  using OptOutDispatch::handle;
+
+ protected:
+  virtual void unhandled(Node* stmt) override;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h]: Same here.

File: torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.cpp [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+    TORCH_INTERNAL_ASSERT(
+        false, "Handle not overriden for ", stmt->getValType().value(), ".");
+  } else {
+    TORCH_INTERNAL_ASSERT("Unrecognized Node type.");
+  }
+}
+
+// Vals
+void OptOutConstDispatch::handle(const IterDomain* stmt) {
+  unhandled(stmt->as<Node>());
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.cpp]: Also here does it need `as<Node>`?

**Review by naoyam — COMMENTED** [review]
Overall looks good to me, just want to confirm some uses of `as<>`.

File: torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp:354]: Yes, was planning to do this in a follow up to try and collapse some redundant code in lowering passes.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Accidental, thanks.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**